doc/dev/adr/0020-error-modeling-headers.md

```diff
@@ -0,0 +1,103 @@
+
+In addition, users must be able to include arbitrary user properties in their command response to support this error reporting in case our standard fields are insufficient. This feature is likely in place already for all languages.
+
+On the command invoker side, we will add APIs for checking if a response was an application error and returning the error code and error data fields parsed from the MQTT message "AppErrCode" and "AppErrPayload" user properties.
+
+Similar to how our SDKs handle serializing the actual MQTT message payload, our SDKs will require the user provide the serializer for serializing/deserializing the AppErrPayload object.
+
+Other than these two new user properties, the over-the-wire behavior of our protocol won't change as a result of this decision.
+
+In order to provide a strongly-typed experience, we will also add codegen support for modeling both the possible error codes (enum-like list of possible values?) and the type of the error payload in DTDL. This modeling will be detailed in a separate ADR, though.
+
+By convention, the value of the AppErrPayload object should be a UTF-8 encoded JSON string since MQTT user property values must be UTF-8 encoded strings.
+
+## Code Example
+
+### Command executor side
+
+```csharp
+public ExtendedResponse<IncrementResponsePayload> Increment(IncrementRequestPayload request)
+{
+    if (request.IncrementValue < 0)
+    {
+        var response = new ExtendedResponse<IncrementResponsePayload>()
+        {
+            Response = new IncrementResponsePayload { CounterResponse = _counter },
+        };
+
+        // Specify error code, but no error payload  
+        response.WithApplicationError("negativeValue");
+
+        // Or you can specify error code and error payload
+        response.WithApplicationError(
+            "negativeValue",
+            new CustomErrorPayload() { ArgumentValue = request.IncrementValue },
+            new ErrorPayloadJsonSerializer(new TestEnvoys.Utf8JsonSerializer()));
+
+        return response;
+    }
+
+    // happy path omitted
+}
+```
+
+### Command invoker side
+
+```csharp
+public void main()
+{
+    MqttSessionClient mqttClient = ...;();
+    CounterClient counterClient = new CounterClient(mqttClient);
+
+    IncrementRequestPayload payload = new IncrementRequestPayload
+    {
+        IncrementValue = -1
+    };
+
+    var response = counterClient.Increment(executorId, payload).WithMetadata();
+    
+    // Check the RPC response for an application error
+    if (response.TryGetApplicationError(SomeJsonSerializer, out string? errorCode, out CustomErrorPayload? errorPayload))
+    {
+        // use the error code and strongly typed custom error payload type as wanted.
+    }
+
+    // Alternatively just fetch the error code if you don't want to deserialize the payload yet.
+    if (response.TryGetApplicationError(out string? errorCode))
```

Comment:
Can we think about using something like: https://dotnet.github.io/dotNext/features/core/result.html That no matter how the "user" models errors (payload or header), there is kind of more opinionated error handling. In the current proposal the customer can just ignore the error and access the payload object

Reply:
> no matter how the "user" models errors (payload or header), there is kind of more opinionated error handling

I'm not sure this is feasible. Partly because a user may not have any error modeling (in payload or headers), and partly because our library allows users to model errors in payload without feeding that information to code gen. We really don't have a way to force users to do all error modeling in a standard way that our library could then parse out.

I think it is reasonable to assert that the user of an RPC invoker will know where to expect where to find an error if one existed (in headers vs in payload), but do you feel differently?

Reply:
I suspect Alex means that if the SDK understands the meaning of the header, it can determine that the response indicates a user error, and it can throw an exception or return an error value instead of returning the normal result value from the payload.  ADR 19 describes how to do this using codegen support, but if the SDKs are directly aware of user-error headers, this could be done without any need for codegen.

However, on reflection, I'm not sure how this would be compatible with [Alex's comment below](https://github.com/Azure/iot-operations-sdks/pull/676#issuecomment-2787512103) that we might need to support both valid response payloads and reported user errors in the same response message.